File: packages/provide-uterm-go/defaults/defaults_test.go
//
// SPDX-FileCopyrightText: Copyright (c) 2025-2026 provide.io llc. All rights reserved.
// SPDX-License-Identifier: AGPL-3.0-or-later
//

package defaults

import (
	"os"
	"path/filepath"
	"testing"
)

func TestConstantValues(t *testing.T) {
	stringConsts := []struct {
		name string
		got  string
		want string
	}{
		{"TelnetHost", TelnetHost, "127.0.0.1"},
		{"BindAll", BindAll, "0.0.0.0"},
		{"ProxyWSPath", ProxyWSPath, "/ws/terminal"},
		{"ServerHost", ServerHost, "127.0.0.1"},
	}
	for _, c := range stringConsts {
		if c.got != c.want {
			t.Errorf("%s = %q, want %q", c.name, c.got, c.want)
		}
	}

	intConsts := []struct {
		name string
		got  int
		want int
	}{
		{"TelnetPort", TelnetPort, 2102},
		{"SSHPort", SSHPort, 2222},
		{"GatewayTelnetPort", GatewayTelnetPort, 2112},
		{"GatewaySSHPort", GatewaySSHPort, 2222},
		{"ProxyPort", ProxyPort, 8765},
		{"ServerPort", ServerPort, 8780},
		{"TelnetRemotePort", TelnetRemotePort, 23},
		{"SSHRemotePort", SSHRemotePort, 22},
		{"WSPingInterval", WSPingInterval, 20},
		{"WSPingTimeout", WSPingTimeout, 20},
		{"WSCloseTimeout", WSCloseTimeout, 10},
		{"ReconnectMaxRetries", ReconnectMaxRetries, 5},
	}
	for _, c := range intConsts {
		if c.got != c.want {
			t.Errorf("%s = %d, want %d", c.name, c.got, c.want)
		}
	}

	if ReconnectBaseBackoffS != 0.5 {
		t.Errorf("ReconnectBaseBackoffS = %v, want 0.5", ReconnectBaseBackoffS)
	}
	if ReconnectMaxBackoffS != 30.0 {
		t.Errorf("ReconnectMaxBackoffS = %v, want 30.0", ReconnectMaxBackoffS)
	}
}

func TestTokenFile(t *testing.T) {
	got, err := TokenFile()
	if err != nil {
		t.Fatalf("TokenFile() error: %v", err)
	}
	home, err := os.UserHomeDir()
	if err != nil {
		t.Fatalf("UserHomeDir() error: %v", err)
	}
	want := filepath.Join(home, ".uterm", "session_token")
	if got != want {
		t.Errorf("TokenFile() = %q, want %q", got, want)
	}
}

func TestTokenFileNoHome(t *testing.T) {
	// os.UserHomeDir fails when $HOME is empty on Unix.
	t.Setenv("HOME", "")
	if _, err := TokenFile(); err == nil {
		t.Error("TokenFile() expected an error when $HOME is unset")
	}
}

File: packages/provide-uterm-ts/src/defaults/defaults.test.ts
//
// SPDX-FileCopyrightText: Copyright (c) 2025-2026 provide.io llc. All rights reserved.
// SPDX-License-Identifier: AGPL-3.0-or-later
//

import { homedir } from "node:os";
import { join } from "node:path";
import { describe, expect, it } from "vitest";
import * as defaults from "./index.ts";

describe("defaults", () => {
  it("matches the TerminalDefaults constants byte-for-byte", () => {
    expect({
      TELNET_HOST: defaults.TELNET_HOST,
      TELNET_PORT: defaults.TELNET_PORT,
      SSH_PORT: defaults.SSH_PORT,
      GATEWAY_TELNET_PORT: defaults.GATEWAY_TELNET_PORT,
      GATEWAY_SSH_PORT: defaults.GATEWAY_SSH_PORT,
      BIND_ALL: defaults.BIND_ALL,
      PROXY_PORT: defaults.PROXY_PORT,
      PROXY_WS_PATH: defaults.PROXY_WS_PATH,
      PROXY_POLL_MS: defaults.PROXY_POLL_MS,
      SERVER_HOST: defaults.SERVER_HOST,
      SERVER_PORT: defaults.SERVER_PORT,
      TELNET_REMOTE_PORT: defaults.TELNET_REMOTE_PORT,
      SSH_REMOTE_PORT: defaults.SSH_REMOTE_PORT,
      WS_PING_INTERVAL: defaults.WS_PING_INTERVAL,
      WS_PING_TIMEOUT: defaults.WS_PING_TIMEOUT,
      WS_CLOSE_TIMEOUT: defaults.WS_CLOSE_TIMEOUT,
      RECONNECT_MAX_RETRIES: defaults.RECONNECT_MAX_RETRIES,
      RECONNECT_BASE_BACKOFF_S: defaults.RECONNECT_BASE_BACKOFF_S,
      RECONNECT_MAX_BACKOFF_S: defaults.RECONNECT_MAX_BACKOFF_S,
    }).toStrictEqual({
      TELNET_HOST: "127.0.0.1",
      TELNET_PORT: 2102,
      SSH_PORT: 2222,
      GATEWAY_TELNET_PORT: 2112,
      GATEWAY_SSH_PORT: 2222,
      BIND_ALL: "0.0.0.0",
      PROXY_PORT: 8765,
      PROXY_WS_PATH: "/ws/terminal",
      PROXY_POLL_MS: 50,
      SERVER_HOST: "127.0.0.1",
      SERVER_PORT: 8780,
      TELNET_REMOTE_PORT: 23,
      SSH_REMOTE_PORT: 22,
      WS_PING_INTERVAL: 20,
      WS_PING_TIMEOUT: 20,
      WS_CLOSE_TIMEOUT: 10,
      RECONNECT_MAX_RETRIES: 5,
      RECONNECT_BASE_BACKOFF_S: 0.5,
      RECONNECT_MAX_BACKOFF_S: 30.0,
    });
  });

  it("resolves the resume-token file under the user home directory", () => {
    expect(defaults.tokenFile()).toBe(join(homedir(), ".uterm", "session_token"));
  });
});
